drivers/w3s.go

```diff
@@ -0,0 +1,341 @@
+package drivers
+
+import (
+	"context"
+	"fmt"
+	bserv "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	"github.com/ipfs/go-unixfs"
+	"github.com/ipld/go-car"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+)
+
+const w3SDefaultSaveTimeout = 2 * time.Minute
+
+var cidV1 = merkledag.V1CidPrefix()
+
+// This represents the main CAR directory structure organized by pubId.
+// Data for each pubId is removed after the CAR directory is published.
+var (
+	dataToPublish   = make(map[string]*rootCar)
+	dataToPublishMu sync.Mutex
+)
+
+type rootCar struct {
+	root    *merkledag.ProtoNode
+	dag     format.DAGService
+	carCids []string
+	mu      sync.Mutex
+}
+
+func newRootCar() *rootCar {
+	return &rootCar{
+		root: newDir(),
+		dag:  merkledag.NewDAGService(bserv.New(blockstore.NewBlockstore(dssync.MutexWrap(ds.NewMapDatastore())), nil)),
+	}
+}
+
+type W3sOS struct {
+	ucanKey   string
+	ucanProof string
+	dirPath   string
+	pubId     string
+}
+
+type W3sSession struct {
+	os *W3sOS
+}
+
+func NewW3sDriver(ucanKey, ucanProof, dirPath, pubId string) *W3sOS {
+	return &W3sOS{
+		ucanKey:   ucanKey,
+		ucanProof: ucanProof,
+		dirPath:   dirPath,
+		pubId:     pubId,
+	}
+}
+
+func (ostore *W3sOS) NewSession(filename string) OSSession {
+	if filename != "" {
+		panic("File names are not supported by W3S driver")
+	}
+	session := &W3sSession{
+		os: ostore,
+	}
+	return session
+}
+
+func (ostore *W3sOS) UriSchemes() []string {
+	return []string{}
+}
+
+func (ostore *W3sOS) Description() string {
+	return "Web3 Storage driver."
+}
+
+func (session *W3sSession) OS() OSDriver {
+	return session.os
+}
+
+func (session *W3sSession) EndSession() {
+	// no op
+}
+
+func (session *W3sSession) ListFiles(ctx context.Context, cid, delim string) (PageInfo, error) {
+	// not supported
+	return nil, nil
+}
+
+func (session *W3sSession) ReadData(ctx context.Context, name string) (*FileInfoReader, error) {
+	// not supported
+	return nil, nil
+}
+
+func (session *W3sSession) IsExternal() bool {
+	return false
```

Comment:
@victorges I saw the comment on the interface `// Indicates whether data may be external to this node` but am still not super clear on what this is intended for

Reply:
To be honest, I'm not sure where it's used. Not in catalyst-api, so I think it's safer to return `false` here.

Reply:
Yeah, not worth spending more time on for this PR but I raised https://github.com/livepeer/go-tools/issues/14 to clean up if possible in the future